### PR TITLE
DS-4287: Helper functions to require a QuestionType

### PR DESCRIPTION
**Pull request by chrisfacer**
The function RequireQuestionType() is to allow Standard R items to enforce that a collection of variables all have the desired questiontype attribute.

### CODE REVIEW

File: R/utilites.R [hunk 1]
```diff
@@ -287,3 +287,79 @@ IsQTable <- function(x)
...
+#'        "Ranking", or "Experiment".
+#' @param message.prefix Text to display in the error message before the 
+#'        required question type / structure.
+#' @param message.suffix Text to display at the end of the message.
+#'
+#' @export
+RequireQuestionType <- function(variables, required.type, message.prefix, message.suffix) {
+    found.question.types <- vapply(variables,
+                             FUN = function(v) { return(attr(v, "questiontype"))},
+                             FUN.VALUE = character(1))
```
Comment (jrwishart) [inline R/utilites.R]: Can make this less verbose with
```suggestion
    found.question.types <- vapply(variables,
                                   FUN = attr,
                                   FUN.VALUE = character(1),
                                   which = "questiontype",
                                   exact = TRUE)
```
The last line forces an exact match

File: R/utilites.R [hunk 2]
```diff
@@ -287,3 +287,79 @@ IsQTable <- function(x)
...
+                                "NumberMulti" = "Numeric - Multi",
+                                "NumberGrid" = "Numeric - Grid",
+                                "PickAnyGrid" = "Binary - Grid",
+                                "Text" = "Text",
+                                "TextMulti" = "Text - Multi", 
+                                "Ranking" = "Ranking",
+                                "Experiment" = "Experiment"))
+    structure.name <- question.types[[product.name]][[type]]
+    if (is.null(structure.name))
+        stop(paste0(type, " is not a valid Question Type to supply."))
```
Comment (jrwishart) [inline R/utilites.R]: `paste0` not required

File: R/utilites.R [hunk 3]
```diff
@@ -287,3 +287,79 @@ IsQTable <- function(x)
...
+#' @param type A string indicating the required questiontype value.
+#'        Should be one of: "PickAny", "PickOne", "PickOneMulti", "Number",
+#'        "NumberMulti", "NumberGrid", "PickAnyGrid", "Text", "TextMulti",
+#'        "Ranking", or "Experiment".
+#' @param product.name A string, which should be either Displayr or Q
+#'
+#' @export
+GetTranslatedQuestionType <- function(type, product.name) {
+    if (! product.name %in% c("Displayr", "Q"))
+        stop("The product name should be either Q or Displayr.")
```
Comment (jrwishart) [inline R/utilites.R]: Could perhaps make use of the `stopifnot` function
```suggestion
    stopifnot("The product name should be either Q or Displayr." = product.name %in% c("Displayr", "Q"))
```

File: R/utilites.R [hunk 4]
```diff
@@ -287,3 +287,79 @@ IsQTable <- function(x)
...
+                             FUN = function(v) { return(attr(v, "questiontype"))},
+                             FUN.VALUE = character(1))
+
+    if (any(found.question.types != required.type)) {
+        # productName is an environment variable present in Q/Displayr
+        product.name <- get0("productName", ifnotfound = "Displayr")
+        structure.name <- GetTranslatedQuestionType(required.type, product.name)
+        structure <- if (product.name == "Q") " question"  else " variable set"
+        message.middle <- paste0(structure.name, structure)
+        stop(paste0(message.prefix, message.middle, message.suffix))
```
Comment (jrwishart) [inline R/utilites.R]: `paste0` not required in `stop`

Comment (chrisfacer) [inline R/utilites.R]:
```suggestion
        stop(message.prefix, message.middle, message.suffix)
```

File: R/utilites.R [hunk 5]
```diff
@@ -287,3 +287,79 @@ IsQTable <- function(x)
...
+                                "NumberMulti" = "Numeric - Multi",
+                                "NumberGrid" = "Numeric - Grid",
+                                "PickAnyGrid" = "Binary - Grid",
+                                "Text" = "Text",
+                                "TextMulti" = "Text - Multi", 
+                                "Ranking" = "Ranking",
+                                "Experiment" = "Experiment"))
+    structure.name <- question.types[[product.name]][[type]]
+    if (is.null(structure.name))
+        stop(paste0(type, " is not a valid Question Type to supply."))
```
Comment (chrisfacer) [inline R/utilites.R]:
```suggestion
        stop(type, " is not a valid Question Type to supply.")
```

File: R/utilites.R [hunk 6]
```diff
@@ -287,3 +287,79 @@ IsQTable <- function(x)
...
+    structure.name
+}
```
Comment (chrisfacer) [inline R/utilites.R]:
```suggestion
}

```

**Pull request merged**